File: 290-word-pattern/word-pattern.java
import java.util.*;

class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");
        if (pattern.length() != words.length) return false;

        Set<Character> chars = new HashSet<>();
        Set<String> uniqWords = new HashSet<>();
        Set<Map.Entry<Character, String>> pairs = new HashSet<>();

        for (int i = 0; i < words.length; i++) {
            char ch = pattern.charAt(i);
            String w = words[i];
            chars.add(ch);
            uniqWords.add(w);
            pairs.add(new AbstractMap.SimpleEntry<>(ch, w));
        }

        return chars.size() == uniqWords.size() && uniqWords.size() == pairs.size();
    }
}

File: 290-word-pattern/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        # words = s.split()
        # if len(pattern) != len(words):
        #     return False

        # p2w = {}
        # w2p = {}

        # for ch, w in zip(pattern, words):
        #     if ch in p2w and p2w[ch] != w:
        #         return False
        #     if w in w2p and w2p[w] != ch:
        #         return False
        #     p2w[ch] = w
        #     w2p[w] = ch

        # return True
        
        words = s.split()
        return (len(pattern) == len(words) and len(set(zip(pattern, words))) == len(set(pattern)) == len(set(words)))

